### PR TITLE
Accessible settings menu in Game Lab and App Lab

### PR DESCRIPTION
**Pull request by bencodeorg**
Refactors menu options from settings dropdown in Applab and Gamelab to use Javalab's dropdown, which is more accessible. Removes usage of React portals, which do not meet accessibility standards because they are rendered at the end of the DOM.

The menu UI changes a little bit (no little caret to point the menu options to the settings cog), but I think we could improve that in the future.

**Navigating to menu / opening dialogs with keyboard (before this change, menu options were not tab navigable)**

tab nav settings cog

**UI change**

| Before | After |
| --- | --- |
| image | image |

## Links

- jira ticket: [SL-373](https://codedotorg.atlassian.net/browse/SL-373)

## Testing story

It looks like the SettingsCog component (and its menu) is used in three places, each of which I confirmed looked as expected:
- Gamelab
- Applab ("Code" selected from Code | Design | Data options)
- Applab ("Design" selected from Code | Design | Data options)

I also looked at a couple of unique states (while run button is clicked, readonly versions of projects, design mode selected in applab, toolbox hidden) and saw that things looked good. I had to fix up the header for "design" mode, which wasn't turning light purple when the run button was clicked ("Data" and "Code" headers turn light purple).

Side note -- I found the header pretty difficult to work with, and kept finding header permutations that I needed to fix. We have a lot of different permutations of headers that look very similar that are rendered completely independently (eg, in Applab, the design workspace and code workspace are rendered completely separately and hidden/shown by the Code | Data | Design button). This is probably desirable in some ways (eg, snappy navigation), but makes it so you have to make styling changes in multiple places.

## Deployment strategy

Should be merged into https://github.com/code-dot-org/code-dot-org/pull/49248, then both changes should be merged together.

## Follow-up work

I'm not sure if the approach I added for `JavalabDropdown` (and reused here) is ideal from an accessibility perspective -- it seems like it might be more appropriate to use `ul` elements for each dropdown option:

https://www.w3.org/WAI/tutorials/menus/application-menus-code/

Didn't want to have to start looking for regressions in Javalab dropdowns as well, but I think this could be a good follow-up.

### CODE REVIEW

File: apps/src/lib/ui/SettingsCog.jsx [hunk 1]
```diff
@@ -117,42 +106,40 @@ export default class SettingsCog extends Component {
...
-          )}
-          {this.areAIToolsEnabled() && (
-            <ManageModels onClick={this.manageModels} />
-          )}
-          {this.props.showMakerToggle && (
-            <ToggleMaker onClick={this.toggleMakerToolkit} />
-          )}
-        </PopUpMenu>
+        {this.state.open && (
+          <JavalabDropdown>
```
Comment (molly-moen) [inline apps/src/lib/ui/SettingsCog.jsx:125]: may be better to do as a follow-up, but should we move and rename `JavalabDropdown` so it is clear it is a reusable component? It's a little confusing to have it here

File: apps/src/StudioApp.js [hunk 2]
```diff
@@ -2443,7 +2443,7 @@ StudioApp.prototype.handleEditCode_ = function(config) {
         this.editor.enablePalette(!this.editor.session.paletteEnabled);
         showToolboxHeader.style.display = this.editor.session.paletteEnabled
           ? 'none'
-          : 'inline-block';
+          : 'flex';
```
Comment (molly-moen) [inline apps/src/StudioApp.js:2446]: more evidence that there is too much is StudioApp!

**Review by molly-moen — APPROVED** [review]
LGTM, one comment that could be a follow-up

File: apps/src/StudioApp.js (hunk 2 above)
Comment (bencodeorg) [reply apps/src/StudioApp.js:2446]: ugh yes, banged my head against a wall for a while before I found this little nugget 🤦

File: apps/src/lib/ui/SettingsCog.jsx (hunk 1 above)
Comment (bencodeorg) [reply apps/src/lib/ui/SettingsCog.jsx:125]: Yeah good point, I think an immediate follow-up makes sense. Any recollection of where we theorized reusable components should go during our apps organization discussion at in-person week?

Comment (molly-moen) [reply apps/src/lib/ui/SettingsCog.jsx:125]: I wasn't in that meeting, but per the [doc](https://docs.google.com/document/d/1IVSopUi7e4Wu_87o2GFZYpTJqs73E755V9Cxq2_3Moc/edit#heading=h.8ii0vz5oblwl) it doesn't look like any conclusions were made?

Comment (bencodeorg) [reply apps/src/lib/ui/SettingsCog.jsx:125]: d'oh, thanks! will review

File: apps/src/StudioApp.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/StudioApp.js:2446]: Could you explain why you had to change from 'inline-block' to 'flex'?

File: apps/src/lib/ui/SettingsCog.jsx [hunk 3]
```diff
@@ -31,10 +33,9 @@ export default class SettingsCog extends Component {
     managingModels: false
   };
 
-  targetPoint = {top: 0, left: 0};
-
   open = () => this.setState({open: true});
   close = () => this.setState({open: false});
+  toggleOpen = () => this.setState({open: !this.state.open});
```
Comment (fisher-alice) [inline apps/src/lib/ui/SettingsCog.jsx:38]: This toggle feature may be nice to add to the Java Lab tab menu.

File: apps/src/lib/ui/SettingsCog.jsx (hunk 1 above)
Comment (fisher-alice) [reply apps/src/lib/ui/SettingsCog.jsx:125]: Thanks for sharing the doc @molly-moen. I enjoyed browsing through [refactor.md](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/docs/refactor.md) that @madelynkasula shared. Wonder if this component would belong in either the mysterious `templates` directory or maybe in `lib/ui` where SettingsCog lives?

File: apps/src/StudioApp.js (hunk 2 above)
Comment (bencodeorg) [reply apps/src/StudioApp.js:2446]: mmm I made the element with this ID (`showToolboxHeader`) a flex container in CodeWorkspace.jsx, and this code changes this attribute when the toolbox is shown/hidden in Gamelab/Applab (not sure why this code is here, to Molly's point). So, this is for consistency with what's happening in CodeWorkspace.jsx.

I don't totally remember why I changed that container to a flexbox -- generally, I like using flex containers when you have 3 elements and want a left, centered, and right element (which is what we want here).

File: apps/src/lib/ui/SettingsCog.jsx (hunk 1 above)
Comment (bencodeorg) [reply apps/src/lib/ui/SettingsCog.jsx:125]: Maybe `apps/src/componentLibrary`? I don't think the JavalabDropdown component is thaaat great right now, but maybe if we put it there we could keep improving it / start using it in more places?

File: apps/src/lib/ui/SettingsCog.jsx (hunk 3 above)
Comment (bencodeorg) [reply apps/src/lib/ui/SettingsCog.jsx:38]: Do you mean this one? Seems like a bug that it doesn't close when I click the "up" arrow:

image

Comment (fisher-alice) [reply apps/src/lib/ui/SettingsCog.jsx:38]: Yes - agree it should close when 'up' is closed.

File: apps/src/lib/ui/SettingsCog.jsx (hunk 1 above)
Comment (fisher-alice) [reply apps/src/lib/ui/SettingsCog.jsx:125]: `apps/src/componentLibrary` would make sense based on naming. We could follow up on the discussion captured by [this doc](https://docs.google.com/document/d/1IVSopUi7e4Wu_87o2GFZYpTJqs73E755V9Cxq2_3Moc/edit#) and provide some kind of guidance/documentation on how to determine which directory a reusable component would be stored.

File: apps/src/StudioApp.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/StudioApp.js:2446]: Thanks so much! Where would you expect to see this snippet of code if not in `StudioApp.js`? `CodeWorkspace.jsx`?

Comment (bencodeorg) [reply apps/src/StudioApp.js:2446]: I think so? I'm sure there's some reason it's there, but it's unusual that we'd have such specific CSS (only applicable to a small part Gamelab and Applab) in here.

**Pull request merged**